sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/AbstractStep.java

```diff
@@ -61,11 +61,14 @@ public Status getStatus() {
      * @param newStatus the new status to be set
      */
     protected void setStatus(Status newStatus) {
-        Status oldStatus = status;
-        status = newStatus;
-        logger.info("{}: changed status from: {} to: {} (interrupted={})",
-                getName(), oldStatus, newStatus, interrupted);
-
+        Status oldStatus;
+        synchronized (statusLock) {
+            oldStatus = status;
+            status = newStatus;
+            logger.info("{}: changed status from: {} to: {} (interrupted={})",
+                    getName(), oldStatus, newStatus, interrupted);
+        }
+        // Just in case, avoid possibility of deadlocks by calling out from outside the lock:
         if (!Objects.equals(oldStatus, newStatus)) {
```

Comment:
Yes. The locking is just to ensure that the Step itself behaves consistently wrt its state.

I'm avoiding locking during the notifyObservers call in case the side effect of that call is the callee coming back and changing the step's status or something (deadlock)